src/Avalonia.Controls/SelectionModel.cs

```diff
@@ -691,8 +704,55 @@ private void SelectRangeImpl(IndexPath start, IndexPath end, bool select)
                         info.ParentNode!.Select(info.Path.GetAt(info.Path.GetSize() - 1), select);
                     }
                 });
+        }
+
+        private void BeginOperation()
+        {
+            if (SelectionChanged != null)
+            {
+                _rootNode.BeginOperation();
+            }
+        }
+
+        private void EndOperation()
+        {
+            static IEnumerable<T>? Concat<T>(IEnumerable<T>? a, IEnumerable<T> b)
+            {
+                return a == null ? b : a.Concat(b);
```

Comment:
Not sure how this works, but would we end up enumerating the items even if the app does not request for any of the properties ?

Reply:
The items should only be enumerated if the user requests one of the properties as `SelectionModelChangeSet` uses `yield`: https://github.com/AvaloniaUI/Avalonia/blob/88794773229e5c9fe93f29672c41b9cbac601a89/src/Avalonia.Controls/SelectionModelChangeSet.cs#L112-L126

As mentioned before however, this could cause a problem if a selection change is made in the `SelectionChanged` handler... We may need to create a new `SelectionModelChangeSet` object for each operation.

Reply:
I've refactored this class, it still only creates the collections on demand but it does actually materialize a collection with all indexes/items. I intend to update it to use the `SelectedItems` collection so that the index/items collections are never actually materialized.